<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Using JBrowse Circular Genome View</title>
    <!-- The commented scripts below are the minified production versions of the
      other scripts -->
    <script src="https://unpkg.com/react@16/umd/react.development.js" crossorigin="anonymous"></script>
    <script src="https://unpkg.com/react-dom@16/umd/react-dom.development.js" crossorigin="anonymous"></script>
    <!--    <script src="https://unpkg.com/@jbrowse/react-circular-genome-view/dist/react-circular-genome-view.umd.development.js"
                crossorigin="anonymous"></script>-->
    <script
            src="https://s3.amazonaws.com/jbrowse.org/jb2_releases/react-circular-genome-view/react-circular-genome-view%40v1.3.2/react-circular-genome-view.umd.development.js"
            crossorigin=""
    ></script>
    <!--    <script src="https://unpkg.com/react@16/umd/react.production.min.js" crossorigin></script>
        <script src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js" crossorigin></script>
        <script src="https://unpkg.com/@jbrowse/react-circular-genome-view/dist/react-circular-genome-view.umd.production.min.js"
                crossorigin></script>-->

</head>
<body>

<div id="jbrowse_circular_genome_view" style="z-index: 2048; position: absolute; top:0; left:0;width: 450px; height: 450px;"></div>

<div id="igvDiv"></div>

<script type="module">

    import {makeDraggable} from '../../node_modules/igv-utils/src/index.js'
    import igv from "../../js";

    const el = document.getElementById('jbrowse_circular_genome_view')
    makeDraggable(el, el)

    // Initialize circular view
    const {createElement} = React
    const {render} = ReactDOM
    const {
        createViewState,
        JBrowseCircularGenomeView,
    } = JBrowseReactCircularGenomeView

    // Keep a copy fo the chords sent to JBrowse until we know how to append to the current set.
    const chords = [];

    const circularState = new createViewState({
        assembly: {
            name: 'forIGV',
            sequence: {
                trackId: 'refSeqTrack',
                type: 'ReferenceSequenceTrack',
                adapter: {
                    type: 'FromConfigSequenceAdapter',
                    features: [],
                },
            },
        },
        tracks: [
            {
                trackId: 'firstTrack',
                assemblyNames: ['forIGV'],
                type: 'VariantTrack',
                adapter: {
                    type: 'FromConfigAdapter',
                    features: chords,
                },
            },
        ],
    })

    render(
        createElement(JBrowseCircularGenomeView, {viewState: circularState}),
        document.getElementById('jbrowse_circular_genome_view'),
    )


    // Start igv.js and create regions
    var options =
        {
            genome: "hg19",
            locus: "chr8:77,706,075-77,707,003",
            tracks: [
                // {
                //     id: "skbr3_vcf",
                //     url: "https://s3.amazonaws.com/igv.org.demo/SKBR3/SKBR3_Illumina_delly.tra.pass.vcf",
                //     type: "variant",
                //     format: "vcf",
                //     name: "Translocations",
                // },
                // {
                //     id: "skbr3_bedpe",
                //     url: "https://s3.amazonaws.com/igv.org.demo/SKBR3/SKBR3_Illumina.delly.tra.pass.bedpe",
                //     type: "interact",
                //     format: "bedpe",
                //     name: "Translocations"
                // },
                {
                    url: "https://s3.amazonaws.com/igv.org.demo/SKBR3/SKBR3_Illumina_delly.tra.pass.bam",
                    indexURL: "https://s3.amazonaws.com/igv.org.demo/SKBR3/SKBR3_Illumina_delly.tra.pass.bam.bai",
                    type: "alignment",
                    format: "bam",
                    name: "Alignments",
                    height: 500,
                    onclick: (features) => {
                        if(features ) {
                            sendPairedAlignmentChord(features);
                        }
                        return true;
                    }
                }
            ]
        };

    var igvDiv = document.getElementById("igvDiv");

    igv.createBrowser(igvDiv, options)
        .then(async function (browser) {
            console.log("Created IGV browser");
            window.igvBrowser = browser;

            const regions = [];
            for (let chrName of browser.genome.wgChromosomeNames) {
                const chr = browser.genome.getChromosome(chrName);
                regions.push(
                    {
                        refName: chr.name.substring(3),
                        uniqueId: chr.name.substring(3),
                        start: 0,
                        end: chr.bpLength,
                    }
                )
            }

            circularState.config.assembly.sequence.adapter.features.set(regions)
            circularState.assemblyManager.removeAssembly(
                circularState.assemblyManager.assemblies[0]
            )
            circularState.assemblyManager.addAssembly(circularState.config.assembly)


            // Send all chords
            // const track = browser.findTracks("id", "skbr3_bedpe")[0];
            // const features = await track.getFeatures("all");
            //
            // const chords = [];
            // let fID = 1;
            // for (let f of features) {
            //     chords.push(
            //         {
            //             uniqueId: fID,
            //             refName: f.chr1,
            //             start: f.start1,
            //             end: f.end1,
            //             mate: {
            //                 refName: f.chr2,
            //                 start: f.start2,
            //                 end: f.end2,
            //             },
            //         })
            //     fID++;
            // }
            // circularState.config.tracks[0].adapter.features.set(chords)
            // circularState.session.view.showTrack(circularState.config.tracks[0].trackId)

            function onChordClick(feature, chordTrack, pluginManager) {
                const f1 = feature.data;
                const f2 = f1.mate;
                const flanking = 100;
                const loci = `${f1.refName}:${f1.start - flanking}-${f1.end + flanking} ${f2.refName}:${f2.start - flanking}-${f2.end + flanking}`
                browser.search(loci);
                console.log(feature)
            }

            circularState.pluginManager.jexl.addFunction('onChordClick', onChordClick)
            circularState.config.tracks[0].displays[0].onChordClick.set(
                'jexl:onChordClick(feature, track, pluginManager)'
            )

        })

    let fID = 0;

    function sendPairedAlignmentChord(features) {


        for (let f of features) {
            const mate = f.mate;
            if (mate && mate.chr && mate.position) {
                chords.push(
                    {
                        uniqueId: fID,
                        refName: f.chr,
                        start: f.start,
                        end: f.end,
                        mate: {
                            refName: mate.chr,
                            start: mate.position - 1,
                            end: mate.position,
                        },
                    })
                fID++;
            }
        }
        circularState.config.tracks[0].adapter.features.set(chords)
        circularState.session.view.showTrack(circularState.config.tracks[0].trackId)
    }


</script>

</body>
</html>
